bioy_pkg/subcommands/classifier.py

```diff
@@ -707,11 +715,29 @@ def action(args):
         assignment_columns += blast_results_columns.tolist()
         blast_results = pd.DataFrame(columns=assignment_columns)
     else:
+
         blast_results_post_len = len(blast_results)
         log.info('{} ({:.0%}) valid hits selected'.format(
             blast_results_post_len,
             blast_results_post_len / blast_results_len))
 
+        if args.best_n_hits > 0:
+            blast_results_len = len(blast_results)
+            query_groups = blast_results.groupby(by=['specimen', 'qseqid'])
+
+            # For each query, determine the mismatches of the Nth best hit
+            mismatch_threshold = query_groups.apply(lambda g: g['mismatch'].nsmallest(args.best_n_hits).iloc[-1])
```

Comment:
I would just define a function that returns what you want in the groupby.apply:

def filter_mismatches(df, thresholh):
  if df['mismatch'].nsmallest(args.best_n_hits).iloc[-1]) >= threshold:
     return df

No you don't need the extra apply at line 732.